### PR TITLE
httpd: support temporary redirect from inside async reply

### PR DESCRIPTION
**Pull request by dotnwat**
- The redirect_exception is hard coded for permanent move status. In Redpanda we also have cases where we need to use temporary redirect, which has the same Location header but different http status. This PR allows the status code to be customized for the redirect_exception.

- Additionally, this PR handles redirect for thrown from future returned from handler. Previous the redirect was only handled when returned from the synchronous code inside handler.

- Adds test for redirect_exception

### CODE REVIEW

File: src/http/routes.cc
Comment (dotnwat) [inline src/http/routes.cc:82]: Comment from previous PR cc @nyh.

> I think it's better to put this code in exception_reply(), which already handles all the other types of official exception, instead of splitting the exception-handling code between here and exception_reply()

https://github.com/scylladb/seastar/pull/1569#discussion_r1145927726

File: src/http/routes.cc
Comment (dotnwat) [inline src/http/routes.cc:104]: Note to reviewers: by removing this redirect exception handling we delegate handling to `exception_reply` which handles `redirect_exception` in exactly the same way. However, in `exception_reply` registered exception handlers (via `routes::register_exception_handler`) take precedence. So in principle this could break some users? It seems low risk, but I'm happy to add it back if there is concern.

**Comment by dotnwat** [comment]
cc @nyh @amnonh. this PR is partially replacing https://github.com/scylladb/seastar/pull/1569 which I left a note on and I think it can be closed.

File: src/http/routes.cc
Comment (nyh) [reply src/http/routes.cc:104]: I think this is fine and even better (if someone set an exception handler for redirect_exception, they wanted it to be used, not ignored). I don't see how any user will want to rely on a situation where if he set this handler explicitly, it will be ignored. But please correct me if I'm wrong and it's somehow more serious.

@amnonh maybe you can review too.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
@amnonh I am planning to merge this, so please review if you want to not merge it yet.

**Comment by tchaikov** [comment]
@scylladb/seastar-maint hello maintainers, could you help merge this change?

**Pull request merged**
